File: Python/1146.py
import sys

def findMaxSum(arr: list[int]) -> int:
    result = arr[0]
    curr = 0
    for x in arr:
        curr = x if curr < 0 else curr + x
        if curr > result:
            result = curr
    return result

data = sys.stdin.read().split()
it = iter(data)
size = int(next(it))
matr = [[int(next(it)) for _ in range(size)] for _ in range(size)]

max_sum = matr[0][0]
for top in range(size):
    arr = [0] * size
    for bottom in range(top, size):
        row_bottom = matr[bottom]
        for i in range(size):
            arr[i] += row_bottom[i]
        max_block_sum = findMaxSum(arr)
        max_sum = max(max_sum, max_block_sum)
print(max_sum)

File: C++/1146.cpp
#include <algorithm>
#include <cstddef>
#include <iostream>
#include <iterator>
#include <vector>

using TArray = std::vector<int>;
using TMatrix = std::vector<TArray>;

int findMaxSum(const TArray& array) {
  int result = array.front();
  int curr_max = 0;
  for (auto x : array) {
    curr_max = std::max(curr_max + x, x);
    result = std::max(result, curr_max);
  }
  return result;
}

int main() {
  std::ios_base::sync_with_stdio(false);
  std::cin.tie(nullptr);

  std::size_t size;
  std::cin >> size;
  TMatrix matr(size, TArray(size));
  for (auto&& row : matr) {
    std::copy_n(std::istream_iterator<int>(std::cin), size, row.begin());
  }

  auto max_sum = matr[0][0];
  TArray column_sums(size);
  for (std::size_t top = 0; top < size; ++top) {
    std::fill(column_sums.begin(), column_sums.end(), 0);
    for (std::size_t bottom = top; bottom < size; ++bottom) {
      std::transform(column_sums.begin(), column_sums.end(),
                     matr[bottom].begin(), column_sums.begin(),
                     std::plus<int>());
      auto max_block_sum = findMaxSum(column_sums);
      max_sum = std::max(max_sum, max_block_sum);
    }
  }
  std::cout << max_sum << '\n';
  return 0;
}
